syntax = "proto3";

package proto;

enum Language {
  INVALID_LANGUAGE = 0;
  CPP = 1;
  C = 2;
  PASCAL = 3;
  PYTHON = 4;
  BASH = 5;
  RUBY = 6;
}

enum ScoreMode {
  INVALID_SCORE_MODE = 0;
  MIN = 1;
  MAX = 2;
  SUM = 3;
}

message Dependency {
  string name = 1;
  string path = 2;
}

message SourceFile {
  string path = 1;
  repeated Dependency deps = 2;
  Language language = 3;
}

message TestCase {
  SourceFile generator = 1;
  repeated string args = 2;
  // TODO(edomora97): add generator runtime dependencies (--fastify)
  SourceFile validator = 3;
  string input_file = 4;
  string output_file = 5;
}

message Subtask {
  ScoreMode score_mode = 1;
  float max_score = 2;
  // TODO this should be a map
  repeated TestCase testcases = 3;
}

message GraderInfo {
  Language for_language = 1;
  repeated Dependency files = 2;
}

message Task {
  string name = 1;
  string title = 2;
  // TODO this should be a map
  repeated Subtask subtasks = 3;
  SourceFile official_solution = 4;
  repeated GraderInfo grader_info = 5;
  SourceFile checker = 6;
  float time_limit = 7;
  int64 memory_limit_kb = 8;
  string input_file = 9;
  string output_file = 10;
}
